### DISCUSSION TITLE
How do locals not interfere with each other?

### DISCUSSION THREAD

**Discussion by TheAtomAnt** [cardillan/mindcode#34] [Q&A] [answered]
for example: 

```
allocate stack in bank1

def T(n)
  if n>0
    print(n)
    T(n-1)
    print(n)
  end
end

T(2)
printflush(message1)
```

in this case, n becomes a local. I would expect the output of this program to be: 2112
but it is not, and it seems like n is being globally altered in the recursive function, which makes sense in Logic. The project seems to go though much effort to push return addresses and parameters on the stack, but locals are not maintained (it seems).

Am I understanding this wrong? Did I write my program above incorrectly to expect 2112 as the output? Although I haven't figured out why... the program actually prints "20" and that's it. If I do T(3) it will print "30" and that's it.

I am interested in understanding this better!

**Comment by francois** [comment, marked as answer]
Ooops! I completely forgot about that! Thank you very much for highlighting this issue.

Your expectations are correct, but the generated code isn't. Locals in functions are not actually "local": they're in the global namespace of Mindustry Logic. As you suspected, "but locals are not maintained". In your case, "n" should be pushed on the stack, and restored afterwards.

I will create an issue for tracking purposes, and fix this in the coming days.

Thank you very much for reporting this!

**Comment by cardillan** [comment]
Note: in the provided code, `T(n-1)` needs to be changed to `T(n - 1)`, since `n-1` is interpreted as a variable name - dashes are valid characters in identifiers because of Mindustry constants like `@titanium-conveyor`. I'm not sure whether it was always like this or not.

This has been solved as part of the March 2023 update and the code nowadays indeed prints 2112, closing.
